✨ Add guard for kflex ctx delete: confirmation for non-KubeFlex contexts

rishi-jat opened the pull request:
Summary :-

This PR implements a safety guard for the kflex ctx delete command to prevent accidental deletion of Kubernetes contexts not managed by KubeFlex.

Changes
	•	Added IsContextManagedByKubeflex function to detect if a context is managed by KubeFlex.
	•	Modified ExecuteCtxDelete to:
	•	Display a warning if the context is not managed by KubeFlex.
	•	Prompt the user for confirmation before deleting non-managed contexts.
	•	Support a --force flag to bypass confirmation for automation.
	•	Updated tests to cover both KubeFlex-managed and non-managed contexts.

closes #392

rxinui commented on cmd/kflex/ctx/delete.go:
```diff
@@ -38,21 +41,47 @@ func CommandDelete() *cobra.Command {
 
 // Execute kflex ctx delete
-func ExecuteCtxDelete(cp common.CP, ctxName string, chattyStatus bool) error {
+func ExecuteCtxDelete(cp common.CP, ctxName string, chattyStatus bool, force bool) error {
```
since `--force` is an optional flag, i would like you to implement the golang option pattern to handle this.

See how i implemented it https://github.com/rishi-jat/kubeflex/blob/21c4b31bd1efad3923707f29c72a991d959d5002/cmd/kflex/common/cp.go#L33C1-L55C2

Also, see https://golang.cafe/blog/golang-functional-options-pattern.html

rxinui commented on cmd/kflex/ctx/delete_test.go:
```diff
@@ -71,11 +71,43 @@ func TestDeleteNonExistentContext(t *testing.T) {
 	}
 	nCtx := len(kconf.Contexts)
-	err = ExecuteCtxDelete(cp, noneCtxName, false)
+	err = ExecuteCtxDelete(cp, noneCtxName, false, false)
```
same comment, see above

rxinui commented on cmd/kflex/ctx/delete_test.go:
```diff
@@ -71,11 +71,43 @@ func TestDeleteNonExistentContext(t *testing.T) {
+
+	// Start test
+	cp := common.NewCP(kubeconfigPath, common.WithName(ctxName))
+	err := ExecuteCtxDelete(cp, ctxName, false, true)
```
same comment, see above

rxinui reviewed (changes requested):
Some changes to do, otherwise good job, thank you for keeping the codebase flow as it is.

rishi-jat commented:
Hi @rxinui, the requested changes are done. Please have a look when you get a chance.

rxinui commented on cmd/kflex/ctx/delete.go:
```diff
@@ -38,21 +54,56 @@ func CommandDelete() *cobra.Command {
 			cp := common.NewCP(kubeconfig)
-			return ExecuteCtxDelete(cp, args[0], chattyStatus)
+
+			var options []DeleteOption
+			if force {
+				options = append(options, WithForce())
+			}
+
+			return ExecuteCtxDelete(cp, args[0], chattyStatus, options...)
```
Very good, thank you

rxinui commented:
/ok-to-test

rxinui commented:
/lgtm

rishi-jat commented:
Hi @pdettori ,
Reason for test failure:
The E2E test was failing because, after renaming a kube context, it lost the KubeFlex metadata. As a result, the delete command prompted for confirmation (since it didn’t recognize the context as KubeFlex-managed), which caused the test to hang in CI.
What I did to fix it:
Updated the context rename logic to correctly preserve KubeFlex extensions/metadata.
Modified the E2E test to use the --force flag for context deletion, preventing interactive prompts in CI.
With these changes, renamed contexts are now properly recognized as KubeFlex-managed, and the E2E tests pass in CI.
Let me know if you need any more details!

rxinui commented on cmd/kflex/ctx/rename.go:
```diff
@@ -62,10 +63,47 @@ func ExecuteCtxRename(cp common.CP, ctxName string, newCtxName string, toSwitch
 		Cluster:  newClusterName,
 		AuthInfo: newAuthInfoAdminName,
 	}
+
+	// Copy extensions from the original context to preserve KubeFlex metadata
+	if originalContext.Extensions != nil {
+		newContext.Extensions = make(map[string]runtime.Object)
+		for key, value := range originalContext.Extensions {
+			// For KubeFlex extensions, we need to update the extension data to reflect the new context name
+			if key == kubeconfig.ExtensionKubeflexKey {
+				// Parse the existing extension
+				runtimeExtension := &kubeconfig.RuntimeKubeflexExtension{}
+				if err := kubeconfig.ConvertRuntimeObjectToRuntimeExtension(value, runtimeExtension); err == nil {
+					// Update the extension data to reflect the new context name
+					if runtimeExtension.Data != nil {
+						// Update the control plane name if it matches the old context name
+						if cpName, exists := runtimeExtension.Data[kubeconfig.ExtensionControlPlaneName]; exists && cpName == ctxName {
+							runtimeExtension.Data[kubeconfig.ExtensionControlPlaneName] = newCtxName
+						}
+						// Update the initial context name if it matches the old context name
+						if initialCtxName, exists := runtimeExtension.Data[kubeconfig.ExtensionInitialContextName]; exists && initialCtxName == ctxName {
+							runtimeExtension.Data[kubeconfig.ExtensionInitialContextName] = newCtxName
+						}
+					}
+					newContext.Extensions[key] = runtimeExtension
+				} else {
+					// If we can't parse the extension, just copy it as-is
+					newContext.Extensions[key] = value
+				}
+			} else {
+				// For non-KubeFlex extensions, copy as-is
+				newContext.Extensions[key] = value
+			}
+		}
+	}
+
+	kconf.Contexts[newCtxName] = newContext
```
if the code has to be changed, the only thing to do is to move from

```
kconf.Contexts[newCtxName] = &api.Context{
		Cluster:  newClusterName,
		AuthInfo: newAuthInfoAdminName,
	}
```

to 
```
kconf.Contexts[newCtxName] = &api.Context{
		Cluster:  newClusterName,
		AuthInfo: newAuthInfoAdminName,
                Extensions: kconf.Contexts[ctxName].Extensions // make sure it is a copy of its value
	}
```

rishi-jat commented:
@rxinui All requested changes are done.
The context rename now only copies the Extensions as-is, with no modification to extension data.
Let me know if anything else is needed—thanks for your guidance!

pdettori commented:
/ok-to-test

rishi-jat commented:
Hi @rxinui,
I’ve added a unit test to ensure that kflex ctx delete prompts a warning and confirmation when deleting a context not managed by KubeFlex, exactly as you requested.

rxinui commented on cmd/kflex/ctx/delete_test.go:
```diff
@@ -17,65 +17,136 @@ limitations under the License.
+	oldStdout := os.Stdout
+	rOut, wOut, _ := os.Pipe()
+	os.Stdout = wOut
+	defer func() { os.Stdout = oldStdout }()
```
up until here, i think it is good. You simulate the prompt and put `n` to cancel the operation, nice job!

However, i would like you to change what follows (see my next comment)

rxinui commented on cmd/kflex/ctx/delete_test.go:
```diff
@@ -17,65 +17,136 @@ limitations under the License.
+	var buf bytes.Buffer
+	buf.ReadFrom(rOut)
+	output := buf.String()
+	if !strings.Contains(output, "Warning: Context") || !strings.Contains(output, "Are you sure you want to delete this context?") {
```
Although, this is correct, it is generally a bad idea to test a text-string. Reason is, kubeflex evolves and we might change the prompt content and forget to update it on the test side. Even if we do not forget, it is a pain to update the test as well.

Instead, I suggest to test the behaviour of cancelling the operation, which is deterministic. @rishi-jat 

CC: @pdettori

rxinui reviewed (changes requested):
Thank you, yet still some improvements to provide. Almost there, we appreciate your commitment.

rishi-jat commented:
@rxinui The test is now updated exactly as you requested:
It only checks that the context is not deleted if the user cancels, without checking any output text.

rxinui reviewed (commented):
/lgtm

@pdettori it's all yours now :)

rishi-jat commented:
Hi @pdettori,  
The latest changes are complete and have `/lgtm` from @rxinui.  
Could you please take a look and `/approve` if everything looks good?

rxinui commented:
@dumb0002 would you mind to run `/ok-to-test` :)

dumb0002 commented:
/ok-to-test

rishi-jat commented:
Hi @rxinui @dumb0002,
All tests are passing  if everything looks good, could you please add /approve? Thanks!

rxinui commented:
/approve

dumb0002 commented:
/approve
